index.bs

```diff
@@ -305,17 +253,22 @@ spec: mimesniff; urlPrefix: https://mimesniff.spec.whatwg.org/#
 
       <p>
         The <dfn for='VideoConfiguration' dict-member>contentType</dfn> member
-        represents the MIME type of the video track.
+        represents the [=MIME type=] of the video track.
       </p>
 
       <p>
         To check if a {{VideoConfiguration}} <var>configuration</var> is a
         <dfn>valid video configuration</dfn>, the following steps MUST be run:
         <ol>
           <li>
-            If <var>configuration</var>'s {{VideoConfiguration/contentType}} is
-            not a <a>valid video MIME type</a>, return <code>false</code> and
-            abort these steps.
+            Let |result| be the result of running the [$Check MIME Type Support$]
+            algorithm with |configuration|'s {{VideoConfiguration/contentType}},
+            the |media| <code>video</code>,
+            and {{MediaEncodingType}} or {{MediaDecodingType}} |encodingOrDecodingType|.
+          </li>
+          <li>
+            If |result| is <code>failure</code> or <code>unsupported</code>,
```

Comment:
Two problems here?

1. The _Check MIME Type Support_  steps don't return `unsupported`, but see my proposed change below to fix the steps.

3. If the result is `unsupported`, IIUC, we shouldn't reject the configuration as invalid.  We should pass through the result to _Create a MediaCapabilitiesDecodingInfo_ to set the `supported` property of the `MediaCapabilitiesDecodingInfo` return value.

Reply:
You're right ... we should only reject if parsing the MIME type fails or the type something other than `video` or `application`. I can try to adjust the algorithms.

Reply:
The difficulty I see is that we currently call **Check MIME Type Support** from **valid video configuration** and **valid audio configuration** (via **valid MediaDecodingConfiguration** and **valid MediaConfiguration**) but we need to use the result from **Check MIME Type Support** in **Create a MediaCapabilitiesDecodingInfo**.

What I'd like to propose is we write a complete set of algorithm steps for `decodingInfo()` and in doing so remove the **valid MediaDecodingConfiguration**, **valid MediaConfiguration**, **valid audio configuration**, **valid video configuration**, etc. concepts (and similar for `encodingInfo()`).

This would make this PR a much bigger rewrite, but I think it would takes us in the right direction. The alternative would be to keep **valid video configuration** etc but then we'd need to split **Check MIME Type Support** into two algorithms: one to check basic validity (does it parse, is its type **audio**, **video** or **application**, etc), the other to check for UA support.

Or am I missing an easier way to solve this?

Reply:
Can we take _Check MIME Type Support_ out of the _valid video configuration_ steps and instead call it in  the _Create a MediaCapabilitiesDecodingInfo_ algorithms?  (Similar for audio, encoding, etc.) 

It would be okay to reject early with a `TypeError` because the input is invalid (i.e. correct properties not set) and later also reject with a `TypeError` because the `contentType` could not be parsed.

Reply:
Sorry for the delay, I'm looking into this...

Reply:
I have updated the PR as we discussed today. I have a few questions and observations:

1. I have added internal slots to `AudioConfiguration` and `VideoConfiguration` to store the parsed MIME type. Is this a valid thing to do, given the configuration object is created by user script?
2. In "Create a MediaCapabilitiesDecodingInfo", if `keySystemConfiguration` exists, this spec calls into EME's [Get Supported Configuration](https://www.w3.org/TR/encrypted-media-2/#get-supported-configuration), then [Get Supported Configuration and Consent](https://www.w3.org/TR/encrypted-media-2/#get-supported-configuration-and-consent), then [Get Supported Capabilities for Audio/Video Type](https://www.w3.org/TR/encrypted-media-2/#get-supported-capabilities-for-audio-video-type) ... which uses "valid media MIME type", changes will be needed there to use "parse a MIME type" (maybe we can use the parsed MIME type in the [[mimeType]] slot). https://github.com/w3c/encrypted-media/issues/511 tracks this.
3. Could the `supported` flag in the `decodingInfo()` or `encodingInfo()` output depend on the combination of audio and video MIME types? At the moment, this PR checks support of each individually and sets supported to true if either is supported. Is this the right thing to do?

Reply:
Also in "valid video configuration" I don't think we now have enough checks. To be consistent with the logic before this PR, it would have to check for <code>video</code> or <code>application</code> type and the parameters. More to do here...

Reply:
1.  I don't see any reason why it's not valid.  Internal slots are intended  to be part of the implementation and not visible to user scripts.  I could call in a webidl expert if we need to. 
2. I will defer to @xhwang-chromium and EME experts to handle that :)
3. I would need to talk to folks more directly involved with the implementation, but from what I can tell, Chrome checks the two values independently.   If we really needed to we could amend the steps to be something like:

> If _audioMimeSupport_ is `supported`, _videoMimeSupport_ is `supported`, and the user agent can decode the media represented by `configuration`, set _supported_ to `true` and `false` otherwise.

Reply:
> Also in "valid video configuration" I don't think we now have enough checks. To be consistent with the logic before this PR, it would have to check for video or application type and the parameters.

Yes, it looks like the checks in [2.1.4. MIME types](https://w3c.github.io/media-capabilities/#mime-type) got dropped.

Reply:
I pushed a commit with most of my suggested changes, feel free to comment/amend as you like.

I wasn't able to resolve this issue though, because the steps in _Check MIME Type Support_ need to be updated.  They can't return _failure_ because the steps that call it don't expect that result; with the current PR, _failure_ would result in a _supported_ return value.

And its not clear to me yet whether the checks in 2.1.4.1 and 2.1.4.2 (in the current spec) belong in _Check MIME Type Support_  (and take the _unsupported_ path) or _check for valid {video,audio} configuration_ (and take the _failure_ path).  Might have to check the WPTs to see what implementations do.

Reply:
The checks for `audio`, `video`, `application` etc. and a "single codec" all take the failure path in WPTs.

https://github.com/web-platform-tests/wpt/blob/f54cd920024da0857fdc5b036f16a7d1fd8792fd/media-capabilities/decodingInfo.any.js#L124

I found one WPT case that checks whether a valid mime type is also valid for WebRTC, and it takes the unsupported path:

https://github.com/web-platform-tests/wpt/blob/f54cd920024da0857fdc5b036f16a7d1fd8792fd/media-capabilities/decodingInfo-webrtc.any.js#L137

Reply:
I'm doing some tests, and finding some inconsistencies with the "imply a codec" checks:
* Chrome rejects if there are parameters other than `codecs`, Firefox resolves. The spec requires implementations to reject
* Chrome resolves on `audio/webm` for audio contentType, Firefox rejects. A `codecs` parameter should be required
* Chrome resolves on `video/webm` for video contentType, Firefox rejects. A `codecs` parameter should be required

I also found a difference in parsing behaviour:
* Chrome rejects on `audio/mpeg;` (note the semicolon), Firefox resolves

Reply:
For the non-webrtc cases, I have updated the PR to reflect the behavior of the majority of browsers in this.

For WebRTC, the current spec is not entirely clear in whether the "single-codecs" logic applies.  I filed #238 to discuss that.  I do not want to block merging on resolving that issue.